dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -0,0 +1,558 @@
+  @GuardedBy("INSTANCE_LOCK")
+  private static final AtomicReference<AlluxioEtcdClient> ALLUXIO_ETCD_CLIENT
+      = new AtomicReference<>();
+  private final AtomicBoolean mConnected = new AtomicBoolean(false);
+  private Client mClient;
+  public final ServiceDiscoveryRecipe mServiceDiscovery;
+  public String[] mEndpoints;
+  private final Closer mCloser = Closer.create();
+  // only watch for children change(add/remove) for given parent path
+  private final ConcurrentHashMap<String, Watch.Watcher> mRegisteredWatchers =
+      new ConcurrentHashMap<>();
+
+  /**
+   * CTOR for AlluxioEtcdClient.
+   * @param conf
+   */
+  public AlluxioEtcdClient(AlluxioConfiguration conf) {
+    String clusterName = conf.getString(PropertyKey.ALLUXIO_CLUSTER_NAME);
+    List<String> endpointsList = conf.getList(PropertyKey.ETCD_ENDPOINTS);
+    mEndpoints = endpointsList.toArray(new String[endpointsList.size()]);
+    mServiceDiscovery = new ServiceDiscoveryRecipe(this, clusterName);
+  }
+
+  /**
+   * Get the singleton instance of AlluxioEtcdClient.
+   * @param conf
+   * @return AlluxioEtcdClient
+   */
+  public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
+    if (ALLUXIO_ETCD_CLIENT.get() == null) {
+      try (LockResource lockResource = new LockResource(INSTANCE_LOCK)) {
+        if (ALLUXIO_ETCD_CLIENT.get() == null) {
+          ALLUXIO_ETCD_CLIENT.set(new AlluxioEtcdClient(conf));
+        }
+      }
+    }
+    return ALLUXIO_ETCD_CLIENT.get();
+  }
+
+  /**
+   * Create jetcd grpc client no forcing.
+   */
+  public void connect() {
+    connect(false);
+  }
+
+  /**
+   * Create jetcd grpc client with choice of force or not.
+   * @param force
+   */
+  public void connect(boolean force) {
+    if (mConnected.get() && !force) {
+      return;
+    }
+    mConnected.set(false);
+    // create client using endpoints
+    Client client = Client.builder().endpoints(mEndpoints)
+        .build();
+    if (mConnected.compareAndSet(false, true)) {
+      mClient = client;
+    }
+  }
+
+  /**
+   * Disconnect.
+   * @throws IOException
+   */
+  public void disconnect() throws IOException {
+    close();
+  }
+
+  /**
+   * Watch for a single path or the change among all children of this path.
+   */
+  enum WatchType {
+    CHILDREN,
+    SINGLE_PATH
+  }
+
+  /**
+   * Lease structure to keep the info about a lease in etcd.
+   */
+  public static class Lease {
+    public long mLeaseId = -1;
+    public long mTtlInSec = -1;
+
+    /**
+     * CTOR for Lease.
+     * @param leaseId
+     * @param ttlInSec
+     */
+    public Lease(long leaseId, long ttlInSec) {
+      mLeaseId = leaseId;
+      mTtlInSec = ttlInSec;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("leaseId", mLeaseId)
+          .add("ttl", mTtlInSec)
+          .toString();
+    }
+  }
+
+  public static final long DEFAULT_LEASE_TTL_IN_SEC = 2L;
```

Comment:
nit: i think we prefer all static final constants to be at the top of class definition

Reply:
thanks for catching